src/kmeans.jl

```diff
@@ -18,74 +24,106 @@ const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
-                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
-                 tol::Real=_kmeans_default_tol,
-                 display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:Real
-
-    m, n = size(X)
-    m2, k = size(centers)
-    m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
+####
+#### Exported kmeans and kmeans! functions
+####
+
+"""
+    kmeans!(centers, X)
+
+Update the current centers `centers` (of size `p x k` where `p` is the dimension and `k` the
+number of centroids) using the samples contained in `X` (of size `p x n` where `n` is the number
+of samples).
+"""
+function kmeans!(centers::AbstractMatrix{<:AbstractFloat}, # in: current centers (p x k)
+                 X::AbstractMatrix{<:Real};                # in: sample matrix (p x n)
+                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing, # in: sample weights (n)
+                 maxiter::Integer=_kmeans_default_maxiter, # in: maximum number of iterations
+                 tol::Real=_kmeans_default_tol,            # in: tolerance of change at convergence
+                 display::Symbol=_kmeans_default_display,  # in: level of display
+                 distance::SemiMetric=SqEuclidean())       # in: function to compute distances
+    p, n = size(X)
+    p2, k = size(centers)
+
+    p == p2 || throw(DimensionMismatch("Inconsistent array dimensions for `X` and `centers`."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
+    if !(weights === nothing)
+        length(weights) == n || throw(DimensionMismatch("Incorrect length of weights."))
+    end
 
     assignments = Vector{Int}(undef, n)
-    costs = Vector{T}(undef, n)
     counts = Vector{Int}(undef, k)
-    cweights = Vector{Float64}(undef, k)
 
-    _kmeans!(X, conv_weights(T, n, weights), centers,
-             assignments, costs, counts, cweights, round(Int, maxiter), tol,
-             display_level(display), distance)
+    # check the types to see wheter the updates: centers[i, cj] += X[i, j] * wj
+    # may occur loss of precision through silent casting
+    update_type = float(weights === nothing ? eltype(X) : promote_type(eltype(X), eltype(weights)))
+    update_type <: eltype(centers) || @warn "The type of the centers update ($update_type) is " *
+                                            "wider than that of the given centers " *
+                                            "($(eltype(centers))). This may incur rounding errors."
+
+    _kmeans!(X, weights, centers, assignments, counts,
+             round(Int, maxiter), tol, display_level(display), distance)
```

Comment:
```suggestion
             Int(maxiter), tol, display_level(display), distance)
```
(If the caller passes 1.5 we should really throw an error.)

Reply:
Also, to reduce the specializations of `_kmeans()`
```suggestion
             round(Int, maxiter), Float64(tol), display_level(display), distance)
```
I think `Float64` should be fine, the comparison is done once per iteration.